### PR TITLE
[WFLY-10640] driver-datasource-class-name and driver-xa-datasource-class-name need to be validated when updating

### PR DESCRIPTION
**Pull request by gaol**
Jira: https://issues.jboss.org/browse/WFLY-10640

Please make sure your PR meets the following requirements:
- [ ] Pull Request title is properly formatted: `[WFLY-XYZ] Subject` or `WFLY-XYZ Subject`
- [ ] Pull Request contains link to the JIRA issue(s)
- [ ] Pull Request contains description of the issue(s)
- [ ] Pull Request does not include fixes for issues other than the main ticket
- [ ] Attached commits represent units of work and are properly formatted

Try to check whether the specified `datasource-class` or `xa-datasource-class` is a valid implementation of `javax.sql.DataSource` or `javax.sql.XADataSource` when adding new jdbc driver or adding new (xa-)datasource.

### CODE REVIEW

**Comment by gaol** [comment]
@maeste could you please review? Thanks. :)

**Review by maeste — APPROVED** [review]

File: connector/src/main/java/org/jboss/as/connector/subsystems/datasources/AbstractDataSourceService.java [hunk 1]
```diff
@@ -364,6 +366,16 @@ public CommonDeployment deploy(ServiceContainer serviceContainer) throws DeployE
 
                 DataSources dataSources = null;
                 if (dataSourceConfig != null) {
+                    String dsClsName = dataSourceConfig.getDataSourceClass();
+                    if (dsClsName != null) {
+                        Class<? extends DataSource> dsCls = driverClassLoader().loadClass(dsClsName).asSubclass(DataSource.class);
```
Comment (kabir) [inline connector/src/main/java/org/jboss/as/connector/subsystems/datasources/AbstractDataSourceService.java]: @gaol what happens if the class does not implement the DataSource interface?

File: connector/src/main/java/org/jboss/as/connector/subsystems/datasources/JdbcDriverAdd.java [hunk 2]
```diff
@@ -107,6 +111,24 @@ protected void performRuntime(OperationContext context, ModelNode operation, Mod
             throw new OperationFailedException(ConnectorLogger.ROOT_LOGGER.failedToLoadModuleDriver(moduleName), e);
         }
 
+        if (dataSourceClassName != null) {
+            Class<? extends DataSource> dsCls;
+            try {
+                dsCls = module.getClassLoader().loadClass(dataSourceClassName).asSubclass(DataSource.class);
```
Comment (kabir) [inline connector/src/main/java/org/jboss/as/connector/subsystems/datasources/JdbcDriverAdd.java]: @gaol what happens if the class does not implement the DataSource interface?

File: connector/src/main/java/org/jboss/as/connector/subsystems/datasources/JdbcDriverAdd.java [hunk 3]
```diff
@@ -107,6 +111,24 @@ protected void performRuntime(OperationContext context, ModelNode operation, Mod
...
+            } catch (ClassNotFoundException e) {
+                throw SUBSYSTEM_DATASOURCES_LOGGER.failedToSetDataSourceClass(driverName, e);
+            }
+            checkDSCls(dsCls, DataSource.class);
+        }
+        if (xaDataSourceClassName != null) {
+            Class<? extends XADataSource> dsCls;
+            try {
+                dsCls = module.getClassLoader().loadClass(xaDataSourceClassName).asSubclass(XADataSource.class);
```
Comment (kabir) [inline connector/src/main/java/org/jboss/as/connector/subsystems/datasources/JdbcDriverAdd.java]: @gaol what happens if the class does not implement the DataSource interface?

File: connector/src/main/java/org/jboss/as/connector/subsystems/datasources/AbstractDataSourceService.java [hunk 4]
```diff
@@ -377,6 +389,16 @@ public CommonDeployment deploy(ServiceContainer serviceContainer) throws DeployE
                     }
                     dataSources = new DatasourcesImpl(Arrays.asList(dataSourceConfig), null, drivers);
                 } else if (xaDataSourceConfig != null) {
+                    String xaDSClsName = xaDataSourceConfig.getXaDataSourceClass();
+                    if (xaDSClsName != null) {
+                        Class<? extends XADataSource> xaDsCls = driverClassLoader().loadClass(xaDSClsName).asSubclass(XADataSource.class);
```
Comment (kabir) [inline connector/src/main/java/org/jboss/as/connector/subsystems/datasources/AbstractDataSourceService.java]: @gaol what happens if the class does not implement the DataSource interface?

File: connector/src/main/java/org/jboss/as/connector/subsystems/datasources/AbstractDataSourceService.java (hunk 1 above)
Comment (gaol) [reply connector/src/main/java/org/jboss/as/connector/subsystems/datasources/AbstractDataSourceService.java]: If the class does not implement the DataSource interface, the data source creation fails as well, but with string `java.lang.ClassCastException` in the error message for the CLI user. I will try to improve the error message by catching this case. :) The same for the below comments :)

**Comment by gaol** [comment]
@kabir updated to catch ClassCastException as well

**Pull request merged**